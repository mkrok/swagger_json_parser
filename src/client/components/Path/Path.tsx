import * as React from 'react';
import Button from '~/client/components/Button';
import { useAppData } from '~/client/context/AppStateProvider';
import {
  ButtonWrapper,
  Header,
  PathButton,
  PathDetails,
  Paragraph,
  ParameterDescription,
  ParameterIn,
  ParameterNameAndType,
  ParameterRequired,
  ResponseDescription,
  ResponseStatusCode,
  Section,
  SectionWrapper,
  Summary,
  Title,
} from './';

const { useEffect, useState } = React;

const Path = ({ info, method }) => {
  const [showingDetails, setShowingDetails] = useState(false);
  const [clicked, setClicked] = useState(false);
  const { appData, setAppData } = useAppData();
  const { canWeShowDetails, waiting } = appData;
  useEffect(() => {
    if (showingDetails && waiting) {
      setShowingDetails(false);
      setAppData({ ...appData, canWeShowDetails: true, waiting: false });
    }
    if (clicked && canWeShowDetails) {
      setShowingDetails(true);
      setClicked(false);
      setAppData({ ...appData, canWeShowDetails: false, waiting: false });
    }
  }, [clicked, waiting]);

  return (
    <>
      <PathButton
        method={!info.deprecated && method}
        onClick={() => {
          setClicked(true);
          if (canWeShowDetails) {
            setShowingDetails(true);
            setClicked(false);
            setAppData({ ...appData, canWeShowDetails: false, waiting: false });
          } else {
            setAppData({ ...appData, waiting: true });
          }
        }}
      >
        {method.toUpperCase()}
      </PathButton>
      {showingDetails && (
        <PathDetails>
          <Title color={!info.deprecated && method}>{`${method.toUpperCase()} ${
            info.deprecated ? ' - (DEPRECATED)' : ''
          }`}</Title>
          <Summary>{info.summary}</Summary>
          <Header>Parameters</Header>
          <SectionWrapper>
            {info.parameters.map((parameter, index) => {
              return (
                <Section key={`parameter-${index + 1}`}>
                  <ParameterNameAndType>
                    {`${parameter.name}: ${parameter.type || ''}`}
                  </ParameterNameAndType>
                  <ParameterRequired>{`${parameter.required ? 'required' : ''}`}</ParameterRequired>
                  <ParameterIn>{`in: ${parameter.in}`}</ParameterIn>
                  <ParameterDescription>{`${
                    parameter.description || 'no description'
                  }`}</ParameterDescription>
                  {info.description && (
                    <ParameterDescription>{info.description}</ParameterDescription>
                  )}
                </Section>
              );
            })}
          </SectionWrapper>
          <Header>Responses</Header>
          <SectionWrapper>
            {Object.keys(info.responses).map((response, index) => {
              return (
                <Section key={`response-${index + 1}`}>
                  <ResponseStatusCode>{response}</ResponseStatusCode>
                  <ResponseDescription>{info.responses[response].description}</ResponseDescription>
                </Section>
              );
            })}
          </SectionWrapper>
          <Paragraph>{`Produces: ${info.produces.join(', ')}`}</Paragraph>
          <ButtonWrapper>
            <Button
              text="Close"
              onClick={() => {
                setShowingDetails(false);
                setAppData({ ...appData, canWeShowDetails: true, waiting: false });
              }}
            />
          </ButtonWrapper>
        </PathDetails>
      )}
    </>
  );
};

export default Path;
