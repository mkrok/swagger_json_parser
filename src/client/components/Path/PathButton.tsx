import styled from 'styled-components';
import { backgroundColors, backgroundColorsOnHover } from '~/client/config/colors';

export const PathButton = styled.button.attrs(props => ({
  method: props.method || 'unknown',
}))`
  width: 5em;
  height: 3em;
  background-color: ${props => backgroundColors[props.method] || 'lightgrey'};
  display: flex;
  flex-direction: column;
  font-size: 1em;
  padding: 5px;
  justify-content: center;
  align-items: center;
  margin-right: 5px;
  margin-bottom: 2px;
  color: black;
  border: none;
  border-radius: 0.3em;
  &:hover {
    background-color: ${props => backgroundColorsOnHover[props.method] || 'grey'};
  }
`;
