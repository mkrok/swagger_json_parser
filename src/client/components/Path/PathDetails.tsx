import styled from 'styled-components';
import { colors } from '~/client/config/colors';

export const PathDetails = styled.div`
  font-size: 1em;
  height: auto;
  width: 300px;
  z-index: 999;
  background-color: #171717;
  border: solid 1px grey;
  border-radius: 0.3em;
  padding: 0.5em;
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  overflow: auto;
  margin: 0.5em;
`;

export const Title = styled.p.attrs(props => ({
  color: props.color || 'unknown',
}))`
  display: inline-block;
  color: ${props => colors[props.color] || 'lightgrey'};
  font-size: 1.5em;
  margin: 0;
  margin-bottom: 0.5em;
  text-align: center;
`;

export const Summary = styled.p`
  display: inline-block;
  color: deepskyblue;
  font-size: 1em;
  margin: 0;
  text-align: center;
  margin-bottom: 1em;
`;

export const SectionWrapper = styled.div`
  border: solid 1px grey;
  margin-bottom: 1em;
`;

export const Section = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  color: grey;
  font-size: 1em;
  background-color: #444;
  padding: 0.5em;
  :nth-child(even) {
    background-color: #333;
  }
`;

export const Header = styled.p`
  display: inline-block;
  color: white;
  font-size: 1em;
  margin-bottom: 0;
  text-align: left;
`;

export const ButtonWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  align-items: center;
  font-size: 1.2em;
  margin-top: 1em;
`;

export const ParameterNameAndType = styled.span`
  color: white;
  margin-right: 0.5em;
  margin-bottom: 0.5em;
`;

export const ParameterRequired = styled.span`
  color: red;
  margin-right: 0.5em;
  margin-bottom: 0.5em;
`;

export const ParameterIn = styled.span`
  color: lightgrey;
  margin-right: 0.5em;
  margin-bottom: 0.5em;
`;

export const ParameterDescription = styled.span`
  color: deepskyblue;
  margin-bottom: 0.5em;
`;

export const ResponseStatusCode = styled.span`
  color: white;
  margin-right: 0.5em;
  margin-bottom: 0.5em;
`;

export const ResponseDescription = styled.span`
  color: deepskyblue;
  margin-bottom: 0.5em;
`;

export const Paragraph = styled.p`
  color: white;
  margin-bottom: 0.5em;
`;
