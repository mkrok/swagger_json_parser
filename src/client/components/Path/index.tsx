export { default } from './Path';
export { PathButton } from './PathButton';
export {
  ButtonWrapper,
  Header,
  Paragraph,
  ParameterDescription,
  ParameterIn,
  ParameterNameAndType,
  ParameterRequired,
  PathDetails,
  ResponseDescription,
  ResponseStatusCode,
  Section,
  SectionWrapper,
  Summary,
  Title,
} from './PathDetails';
