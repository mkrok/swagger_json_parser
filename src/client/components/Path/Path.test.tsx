import renderer from 'react-test-renderer';
import Path from './';

describe('Path', () => {
  it('should match snapshot', () => {
    const props = {
      method: 'get',
      info: {},
    };
    const component = renderer.create(<Path {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
