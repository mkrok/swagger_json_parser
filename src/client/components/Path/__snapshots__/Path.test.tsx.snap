// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Path should match snapshot 1`] = `
<button
  className="sc-gueYoa hCHzdx"
  disabled={true}
  method="get"
  onClick={[Function]}
>
  GET
</button>
`;
